feature: first installation (#70)
ergonode/docs: docs/installation.md

# Installation

> **Ergonode** is headless system use two different application [**frontend**][frontend] and [**backend**][backend]. 

Reviewer: ```suggestion
**Ergonode** is a headless system which uses two different application [**frontend**][frontend] and [**backend**][backend]. 
```

> **Ergonode** is a headless system which uses two different application [**frontend**][frontend] and [**backend**][backend]. 

Reviewer 2: Maybe we should add something like this at the beginning: 
"This is the initial installation that allows you to run the application quickly to familiarize yourself with it. The installation runs the application in production mode so to contribute to the project see <link>. "

![Service Architecture](./images/service_architecture.png "Architecture")

> In user site Backend is PHP application which the frontend communicates with using the REST API. 

Reviewer: ```suggestion
From the user site Backend is PHP application to which frontend communicates using the REST API. 
```

## Before installation
* PostgreSQL
* RabbitMQ
* PHP 7.4
* Node 12 or higher
* Composer
* npm

## Setup
For Ergonode to start correctly, it is best to run the applications in the correct order.

#### 1. PostgreSQL
Start or [install PostgreSQL][postgresql-install], of course you can use an external service.

#### 2. RabbitMQ
Start or [install RabbitMQ][rabbitmq-install], of course you can use an external service.

#### 3. Backend application setup.
If you want to install Ergonode on your local machine, you can use following command:
```shell
composer create-project ergonode/skeleton my_backend_project_name
```
It will create a `my_backend_project_name` directory with a brand new backend Ergonode application inside.

> Go to `my_backend_project_name` directory and create ```.env.local``` file and add lines with database access configuration

Reviewer: ```suggestion
Go to `my_backend_project_name` directory, create ```.env.local``` file and add lines with database access configuration
```

```
DATABASE_URL=pgsql://db_user:db_password@127.0.0.1:5432/db_name
```
Generate keys for Json Web Token
```
openssl genrsa -out config/jwt/private.pem -aes256 4096
openssl rsa -pubout -in config/jwt/private.pem -out config/jwt/public.pem
```
While executing above commands you will be asked for your password. This password needs to be saved then in ```.env.local``` file
in line `JWT_PASSPHRASE=yourpassword`

Execute the following commands to initialise the database
```
bin/console doctrine:database:create
bin/console ergonode:migrations:migrate
```

> Fill database with initial data execute command:

Reviewer: ```suggestion
Fill database with initial data executing command:
```

```
It will create a `my_frontend_project_name` directory with a brand new frontend Ergonode application inside.

Go to `my_frontend_project_name` directory install project dependencies:

```bash
npm install
```

Set your local `.env` file:

```bash
npm run env
```

Automatically setup all modules by executing command:
```bash
npm run modules:all
```

Build and start application 

```bash
npm run build
npm run start
```

See more about: [frontend installation][frontend-development]

#### 5. Launching application
* Launching backend php server - easiest [Symfony server][symfony-server] or via [nginx][nginx-symfony-server].
* Lunch node server 

To use the application in your browser, open:
```
localhost:3000
```

If you want to check the operation of the api, backend is:
```
localhost:8000
```

## Contribution

If you are a developer who is interested in **open-source** projects, you're in the right place because **Ergonode** is such a project.<br>
You will be very interested in installing the application this way:
* [Frontend contribution installation][frontend-contribution]
* [Backend contribution installation][backend-contribution]

> We also have for you a docker repository with our project, which will certainly make the work easier for many people:

Reviewer: ```suggestion
We also have a docker repository in our project, which will certainly make the work easier for many people:
```
